File: en/codes/typescript/chapter_graph/graph_bfs.ts
/**
 * File: graph_bfs.ts
 * Created Time: 2023-02-21
 * Author: Zhuo Qinyue (1403450829@qq.com)
 */

import { GraphAdjList } from './graph_adjacency_list';
import { Vertex } from '../modules/Vertex';

/* Breadth-first traversal */
// Use adjacency list to represent the graph, to obtain all adjacent vertices of a specified vertex
function graphBFS(graph: GraphAdjList, startVet: Vertex): Vertex[] {
    // Vertex traversal sequence
    const res: Vertex[] = [];
    // Hash set, used to record visited vertices
    const visited: Set<Vertex> = new Set();
    visited.add(startVet);
    // Queue used to implement BFS
    const que = [startVet];
    // Starting from vertex vet, loop until all vertices are visited
    while (que.length) {
        const vet = que.shift(); // Dequeue the vertex at the head of the queue
        res.push(vet); // Record visited vertex
        // Traverse all adjacent vertices of that vertex
        for (const adjVet of graph.adjList.get(vet) ?? []) {
            if (visited.has(adjVet)) {
                continue; // Skip already visited vertices
            }
            que.push(adjVet); // Only enqueue if not visited
            visited.add(adjVet); // Mark the vertex as visited
        }
    }
    // Return the vertex traversal sequence
    return res;
}

/* Driver Code */
/* Initialize undirected graph */
const v = Vertex.valsToVets([0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
const edges = [
    [v[0], v[1]],
    [v[0], v[3]],
    [v[1], v[2]],
    [v[1], v[4]],
    [v[2], v[5]],
    [v[3], v[4]],
    [v[3], v[6]],
    [v[4], v[5]],
    [v[4], v[7]],
    [v[5], v[8]],
    [v[6], v[7]],
    [v[7], v[8]],
];
const graph = new GraphAdjList(edges);
console.log('\nAfter initialization, the graph is');
graph.print();

/* Breadth-first traversal */
const res = graphBFS(graph, v[0]);
console.log('\nBreadth-first traversal (BFS) vertex sequence is');
console.log(Vertex.vetsToVals(res));

File: en/codes/kotlin/chapter_graph/graph_bfs.kt
/**
 * File: graph_bfs.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_graph

import utils.Vertex
import java.util.*

/* Breadth-first traversal */
// Use adjacency list to represent the graph, to obtain all adjacent vertices of a specified vertex
fun graphBFS(graph: GraphAdjList, startVet: Vertex): MutableList<Vertex?> {
    // Vertex traversal sequence
    val res = mutableListOf<Vertex?>()
    // Hash set, used to record visited vertices
    val visited = HashSet<Vertex>()
    visited.add(startVet)
    // Queue used to implement BFS
    val que = LinkedList<Vertex>()
    que.offer(startVet)
    // Starting from vertex vet, loop until all vertices are visited
    while (!que.isEmpty()) {
        val vet = que.poll() // Dequeue the vertex at the head of the queue
        res.add(vet)         // Record visited vertex
        // Traverse all adjacent vertices of that vertex
        for (adjVet in graph.adjList[vet]!!) {
            if (visited.contains(adjVet))
                continue        // Skip already visited vertices
            que.offer(adjVet)   // Only enqueue unvisited vertices
            visited.add(adjVet) // Mark the vertex as visited
        }
    }
    // Return the vertex traversal sequence
    return res
}

/* Driver Code */
fun main() {
    /* Initialize undirected graph */
    val v = Vertex.valsToVets(intArrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9))
    val edges = arrayOf(
        arrayOf(v[0], v[1]),
        arrayOf(v[0], v[3]),
        arrayOf(v[1], v[2]),
        arrayOf(v[1], v[4]),
        arrayOf(v[2], v[5]),
        arrayOf(v[3], v[4]),
        arrayOf(v[3], v[6]),
        arrayOf(v[4], v[5]),
        arrayOf(v[4], v[7]),
        arrayOf(v[5], v[8]),
        arrayOf(v[6], v[7]),
        arrayOf(v[7], v[8])
    )
    val graph = GraphAdjList(edges)
    println("\nAfter initialization, the graph is")
    graph.print()

    /* Breadth-first traversal */
    val res = graphBFS(graph, v[0]!!)
    println("\nBreadth-first traversal (BFS) vertex sequence is")
    println(Vertex.vetsToVals(res))
}